File: Ruby/lisp.rb
#!/usr/bin/env ruby

require 'readline'

class Env < Hash
  def initialize parms=[], args=[], outer=nil
    parms.zip(args).each do |k,v|
      self[k] = v
    end
    @outer = outer
  end

  attr_reader :outer

  def find var
    return has_key?(var) ? self : @outer.find(var)
  end
end

def infix label, sym=nil
  sym = :"#{label}" if sym == nil
  {label => lambda {|x,y| x.method(sym).call(y) }}
end

def add_globals env
  Math.methods.each do |i| 
    env.update({i.to_s=>Math.method(i)})
  end
  [infix("+") , infix("-") , infix("*") , infix("/") ,
    infix(">") , infix("<") , infix(">=") , infix("<=") ,
    infix("=", :==) , {"not"=>lambda {|x| not x}} ,
    infix("append", :+), infix("equal?", :==) ,
    {"eq?"=>lambda {|x,y| x.__id__ == y.__id__}} ,
    {"length"=>lambda {|x| x.length}} ,
    {"cons"=>lambda {|x,y| [x] + y}} ,
    {"car"=>lambda {|x| x[0]}} ,
    {"cdr"=>lambda {|x| x[1..-1]}} ,
    {"list"=>lambda {|*x| x}} ,
    {"list?"=>lambda {|x| x.class == Array}} ,
    {"null?"=>lambda {|x| x == []}} ,
    {"symbol?"=>lambda {|x| x.class == String}}
  ].each do |h|
    env.update(h)
  end
  return env
end

Global_env = add_globals(Env.new())

# --------------------------------------------------------------

def eval_sexp x , env=Global_env
  if x.class == String
    return env.find(x)[x]
  elsif x.class != Array
    return x
  elsif x[0] == "quote"
    exp = x[1]
    return exp
  elsif x[0] == "if"
    test = x[1]
    conseq = x[2]
    alt = x[3]
    return eval_sexp(eval_sexp(test, env) ? conseq : alt, env)
  elsif x[0] == "set!"
    var = x[1]
    exp = x[2]
    env.find(var)[var] = eval_sexp(exp,env)
  elsif x[0] == "define"
    var = x[1]
    exp = x[2]
    env[var] = eval_sexp(exp,env)
  elsif x[0] == "lambda"
    vars = x[1]
    exp = x[2]
    return lambda {|*args| eval_sexp(exp, Env.new(vars, args, env))}
  elsif x[0] == "begin"
    ret = nil
    x[1..-1].each do |e|
      ret = eval_sexp(e, env)
    end
    return ret
  else
    exps = x.map{|i| eval_sexp(i,env)}
    prc = exps.shift
    return prc.call(*exps)
  end
end

# --------------------------------------------------------------

def read_from tokens
  raise SyntaxError.new("unexpected EOF while reading") if tokens.length == 0
  token = tokens.shift
  if "(" == token
    ret = []
    while tokens.first != ")"
      ret << read_from(tokens)
    end
    tokens.shift
    return ret
  elsif ")" == token
    raise SyntaxError.new("unexpected )")
  else
    return atom(token)
  end
end

def tokenize s
  s.gsub(/\(/," ( ").gsub(/\)/," ) ").split
end

def read s
  read_from(tokenize(s))
end

def parse s
  read s
end

def atom token
  begin
    return Integer(token)
  rescue ArgumentError
    begin
      return Float(token)
    rescue
      return String(token)
    end
  end
end

def to_string exp
	if exp.class == Array
	 	"(#{exp.map{|i| to_string(i)}.join(" ")})"
	else
		exp.to_s
	end
end


def diff_bracket buf
	buf.count("(") - buf.count(")")
end

def read_sentence prompt
	buf = Readline.readline(prompt, true)
	if buf == nil || buf.length == 0
		return buf
	else
		while diff_bracket(buf) > 0
			buf += Readline.readline("", true)
		end

		if diff_bracket(buf) == 0
			return buf
		else # "(" < ")"
			puts "invalid"
			return ""
		end
	end
end

def repl prompt="#{__FILE__}> "
  #while buf = Readline.readline(prompt, true)
  while buf = read_sentence(prompt)
		if buf == ""
			next
		end
    val = eval_sexp(parse(buf))
    puts "-> #{to_string(val)}" unless val == nil
  end
end


if $0 == __FILE__
  repl()
end

File: Go/lisp.go
package main

// https://gist.github.com/ofan/721464

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var GLOBAL_ENV = NewGlobalEnvironment()
var NIL = NewAtom("nil")
var TRUE = NewAtom("#t")
var FALSE = NewAtom("#f")

type Environment struct {
	dictionary map[string]*Cell
	outer      *Environment
}

func NewEnvironment(params []*Cell, args []*Cell, outer *Environment) *Environment {
	m := map[string]*Cell{}
	for i, c := range params {
		m[c.value] = args[i]
	}
	return &Environment{m, outer}
}

func NewGlobalEnvironment() *Environment {
	m := map[string]*Cell{}

	m[NIL.value] = NIL
	m[TRUE.value] = TRUE
	m[FALSE.value] = FALSE

	m["append"] = NewProc(func(a *Cell) *Cell {
		ret := a.list[0]
		for i := 1; i < len(a.list); i++ {
			for _, c := range a.list[i].list {
				ret.Append(c)
			}
		}
		return ret
	})
	m["car"] = NewProc(func(a *Cell) *Cell {
		return a.list[0].list[0]
	})
	m["cdr"] = NewProc(func(a *Cell) *Cell {
		ret := NewList()
		list := a.list[0].list
		for i := 1; i < len(list); i++ {
			ret.Append(list[i])
		}
		return ret
	})
	m["cons"] = NewProc(func(a *Cell) *Cell {
		ret := NewList()
		ret.Append(a.list[0])
		for _, c := range a.list[1].list {
			ret.Append(c)
		}
		return ret
		return nil
	})
	m["length"] = NewProc(func(a *Cell) *Cell {
		return NewAtom(strconv.Itoa(len(a.list[0].list)))
	})
	m["list"] = NewProc(func(a *Cell) *Cell {
		ret := NewList()
		for _, c := range a.list {
			ret.Append(c)
		}
		return ret
	})
	m["list?"] = NewProc(func(a *Cell) *Cell {
		return NewBool(a.list[0].is(TYPE_LIST))
	})
	m["symbol?"] = NewProc(func(a *Cell) *Cell {
		return NewBool(a.list[0].is(TYPE_SYMBOL))
	})
	m["null?"] = NewProc(func(a *Cell) *Cell {
		return NewBool(a.list[0] == NIL)
	})
	m["not"] = NewProc(func(a *Cell) *Cell {
		return NewBool(a.list[0].Equal(a.list[1]))
	})

	m["+"] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for i := 1; i < len(a.list); i++ {
			n += a.list[i].toInt()
		}
		return NewNumber(n)
	})

	m["-"] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for i := 1; i < len(a.list); i++ {
			n -= a.list[i].toInt()
		}
		return NewNumber(n)
	})
	m["*"] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for i := 1; i < len(a.list); i++ {
			n *= a.list[i].toInt()
		}
		return NewNumber(n)
	})
	m["/"] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for i := 1; i < len(a.list); i++ {
			n /= a.list[i].toInt()
		}
		return NewNumber(n)
	})
	m[">"] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for _, v := range a.list[1 : len(a.list)-1] {
			if false == (n > v.toInt()) {
				return NewBool(false)
			}
		}
		return NewBool(true)
	})
	m["<"] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for _, v := range a.list[1 : len(a.list)-1] {
			if false == (n < v.toInt()) {
				return NewBool(false)
			}
		}
		return NewBool(true)
	})
	m["<="] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for _, v := range a.list[1 : len(a.list)-1] {
			if false == (n <= v.toInt()) {
				return NewBool(false)
			}
		}
		return NewBool(true)
	})
	m[">="] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for _, v := range a.list[1 : len(a.list)-1] {
			if false == (n >= v.toInt()) {
				return NewBool(false)
			}
		}
		return NewBool(true)
	})
	m["="] = NewProc(func(a *Cell) *Cell {
		n := a.list[0].toInt()
		for _, v := range a.list[1 : len(a.list)-1] {
			if false == (n == v.toInt()) {
				return NewBool(false)
			}
		}
		return NewBool(true)
	})

	return &Environment{m, nil}
}

func NewBool(cond bool) *Cell {
	if cond {
		return TRUE
	} else {
		return FALSE
	}
}

func (e *Environment) Find(name string) *Environment {
	atom := e.dictionary[name]
	if atom != nil {
		return e
	} else if e.outer != nil {
		return e.outer.Find(name)
	} else {
		return nil
	}
}

func Eval(cell *Cell, env *Environment) *Cell {
	if cell.is(TYPE_SYMBOL) {
		return env.Find(cell.value).dictionary[cell.value]
	} else if cell.is(TYPE_NUMBER) {
		return cell
	} else if len(cell.list) == 0 {
		return NIL
	} else if cell.list[0].is(TYPE_SYMBOL) {
		head := cell.list[0].value
		if head == "quote" {
			return cell.list[1]
		} else if head == "if" {
			if Eval(cell.list[1], env).value == FALSE.value {
				if len(cell.list) < 4 {
					return NIL
				} else {
					return Eval(cell.list[3], env)
				}
			} else {
				return Eval(cell.list[2], env)
			}
		} else if head == "set!" {
			c := Eval(cell.list[2], env)

			key := cell.list[1].value
			e := env.Find(key)
			if e == nil {
				e = env
			}
			e.dictionary[key] = c
			return c
		} else if head == "define" {
			c := Eval(cell.list[2], env)
			key := cell.list[1].value
			env.dictionary[key] = c
			return c
		} else if head == "lambda" {
			cell.cellType = TYPE_LAMBDA
			cell.env = env
			return cell
		} else if head == "begin" {
			for i := 1; i < len(cell.list)-1; i++ {
				Eval(cell.list[i], env)
			}
			return Eval(cell.list[len(cell.list)-1], env)
		}
	}

	proc := Eval(cell.list[0], env)
	exps := NewList()
	for i := 1; i < len(cell.list); i++ {
		exps.Append(Eval(cell.list[i], env))
	}

	if proc.cellType == TYPE_LAMBDA {
		return Eval(proc.list[2], NewEnvironment(proc.list[1].list, exps.list, env))
	} else if proc.cellType == TYPE_PROC {
		return proc.proc(exps)
	} else {
		println("not a function")
		return NIL
	}
}

const (
	TYPE_SYMBOL = iota + 1
	TYPE_NUMBER
	TYPE_LIST
	TYPE_PROC
	TYPE_LAMBDA
)

type Cell struct {
	cellType int
	value    string
	list     []*Cell
	proc     func(*Cell) *Cell

	env *Environment
}

func NewAtom(v string) *Cell {
	c := &Cell{TYPE_SYMBOL, v, nil, nil, nil}
	_, err := strconv.Atoi(v)
	if err == nil {
		c.cellType = TYPE_NUMBER
	}
	return c
}

func NewNumber(num int) *Cell {
	return &Cell{TYPE_NUMBER, strconv.Itoa(num), nil, nil, nil}
}

func NewList() *Cell {
	return &Cell{TYPE_LIST, "", []*Cell{}, nil, nil}
}

func NewProc(proc func(*Cell) *Cell) *Cell {
	return &Cell{TYPE_PROC, "", nil, proc, nil}
}

func (c *Cell) is(typeId int) bool {
	return c.cellType == typeId
}

func (c *Cell) Append(a *Cell) {
	c.list = append(c.list, a)
}

func (c *Cell) toInt() int {
	i, err := strconv.Atoi(c.value)
	if err != nil {
		panic(err)
	}
	return i
}

func (c *Cell) Equal(a *Cell) bool {
	return c.ToString() == a.ToString()
}

func (c *Cell) ToString() string {
	switch c.cellType {
	case TYPE_LIST:
		s := "("
		for i := 0; i < len(c.list); i++ {
			s += " " + c.list[i].ToString()
		}
		s += " )"
		return s
	case TYPE_LAMBDA:
		return "<Lambda>"
	case TYPE_PROC:
		return "<Proc>"
	default:
		return c.value
	}
}

type TokenBuffer struct {
	tokens []string
	index  int
}

func NewTokenBuffer(tokens []string) *TokenBuffer {
	return &TokenBuffer{tokens, 0}
}

func (buf *TokenBuffer) Shift() string {
	ret := buf.tokens[buf.index]
	buf.index++
	return ret
}

func (buf *TokenBuffer) Get() string {
	return buf.tokens[buf.index]
}

func Tokenize(s string) *TokenBuffer {
	pairs := [][]string{
		{`\(`, " ( "},
		{`\)`, " ) "},
		{` +`, " "},
		{` +$`, ""},
		{`^ +`, ""},
	}

	ret := s
	for _, e := range pairs {
		ret = regexp.MustCompile(e[0]).ReplaceAllString(ret, e[1])
	}
	return NewTokenBuffer(strings.Split(ret, " "))
}

func ReadFrom(buf *TokenBuffer) *Cell {
	if len(buf.tokens) == 0 {
		panic("unexpected EOF while reading")
	}

	token := buf.Shift()

	if "(" == token {
		ret := NewList()
		for buf.Get() != ")" {
			ret.Append(ReadFrom(buf))
		}
		buf.Shift()
		return ret
	} else if ")" == token {
		panic("unexpected )")
	} else {
		return NewAtom(token)
	}
}

func Parse(s string) *Cell {
	return ReadFrom(Tokenize(s))
}

func main() {

	const (
		Prompt     = "lisp.go> "
		PromptNext = "-> "
	)

	scanner := bufio.NewScanner(os.Stdin)
	fmt.Print(Prompt)
	for scanner.Scan() {
		s := scanner.Text()
		if len(s) > 0 {
			fmt.Println(Eval(Parse(s), GLOBAL_ENV).ToString())
			fmt.Print(Prompt)
		}
	}

}
